manta-trusted-setup/src/groth16/ceremony/server.rs

```diff
@@ -0,0 +1,169 @@
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Trusted Setup Server
+
+use crate::{
+    ceremony::{
+        registry::Registry,
+        signature::SignedMessage,
+        util::{deserialize_from_file, serialize_into_file},
+    },
+    groth16::{
+        ceremony::{
+            coordinator::Coordinator,
+            message::{CeremonySize, ContributeRequest, QueryRequest, QueryResponse},
+            Ceremony, CeremonyError, Participant as _, UnexpectedError,
+        },
+        mpc::{State, StateSize},
+    },
+};
+use alloc::sync::Arc;
+use core::ops::Deref;
+use manta_util::{
+    serde::{de::DeserializeOwned, Serialize},
+    BoxArray,
+};
+use parking_lot::Mutex;
+use std::{
+    fs::OpenOptions,
+    path::{Path, PathBuf},
+};
+
+/// Server
+pub struct Server<C, R, const LEVEL_COUNT: usize, const CIRCUIT_COUNT: usize>
+where
+    C: Ceremony,
+    R: Registry<C::Identifier, C::Participant>,
+{
+    /// Coordinator
+    coordinator: Arc<Mutex<Coordinator<C, R, CIRCUIT_COUNT, LEVEL_COUNT>>>,
+
+    /// Recovery directory path
+    recovery_directory: PathBuf,
+}
+
+impl<C, R, const LEVEL_COUNT: usize, const CIRCUIT_COUNT: usize>
+    Server<C, R, LEVEL_COUNT, CIRCUIT_COUNT>
+where
+    C: Ceremony,
+    R: Registry<C::Identifier, C::Participant>,
+{
+    /// Builds a ['Server`] with initial `state`, `challenge`, a loaded `registry`, and a
+    /// `recovery_directory`.
+    #[inline]
+    pub fn new(
+        state: BoxArray<State<C>, CIRCUIT_COUNT>,
+        challenge: BoxArray<C::Challenge, CIRCUIT_COUNT>,
+        registry: R,
+        recovery_directory: PathBuf,
+        size: BoxArray<StateSize, CIRCUIT_COUNT>,
+    ) -> Self {
+        Self {
+            coordinator: Arc::new(Mutex::new(Coordinator::new(
+                registry, state, challenge, size,
+            ))),
+            recovery_directory,
+        }
+    }
+
+    /// Recovers from a disk file at `path` and use `recovery_directory` as the backup directory.
+    #[inline]
+    pub fn recover<P>(path: P, recovery_directory: PathBuf) -> Result<Self, CeremonyError<C>>
+    where
+        P: AsRef<Path>,
+        Coordinator<C, R, CIRCUIT_COUNT, LEVEL_COUNT>: DeserializeOwned,
+    {
+        Ok(Self {
+            coordinator: Arc::new(Mutex::new(
+                deserialize_from_file(path)
+                    .map_err(|_| CeremonyError::Unexpected(UnexpectedError::Serialization))?,
+            )),
+            recovery_directory,
+        })
+    }
+
+    /// Gets the server state size and the current nonce of the participant.
+    #[inline]
+    pub async fn start(
```

Comment:
What do we use this function for?

Reply:
This is to give the client the starting state of the ceremony.